crates/ide_completion/src/context.rs

```diff
@@ -290,23 +292,114 @@ impl<'a> CompletionContext<'a> {
+                            // like to use it, when the user has not yet typed
+                            // any part of the struct field expression:
+                            //
+                            // ```
+                            // struct Foo { a: u32 }
+                            // fn foo() { Foo { a: $0 } }
+                            // ```
+                            //
+                            // The trick we are using here is in this case the
+                            // users cursor will be associated with a whitespace
+                            // token which is a sibling to the ast::RecordExprField.
+                            //
+                            // The failure happens because resolve_record_field expects
+                            // the `expr` of that field to be `Some`, which it is not.
+                            //
+                            // Is there some way to directly look up the type of the
+                            // NameRef `a`?
+                            self.token.prev_sibling_or_token()
+                                .map(|se| se.into_node())
+                                .flatten()
+                                .map(|node| ast::RecordExprField::cast(node))
+                                .flatten()
+                                .map(|rf| self.sema.resolve_record_field(&rf))
+                                .flatten()
+                                .map(|f|(
+                                    Some(f.0.signature_ty(self.db)),
+                                    Some(f.0.name(self.db).to_string()),
+                                ))
+                                .unwrap_or((None, None))
+                        },
+                        ast::RecordExprField(it) => {
+                            cov_mark::hit!(expected_type_struct_field_with_leading_char);
+                            self.sema
+                                .resolve_record_field(&it)
+                                .map(|f|(
+                                    Some(f.0.signature_ty(self.db)),
+                                    Some(f.0.name(self.db).to_string()),
+                                ))
+                                .unwrap_or((None, None))
+                        },
+                        ast::MatchExpr(it) => {
+                            cov_mark::hit!(expected_type_match_arm_without_leading_char);
+                            let ty = it.expr()
+                                .map(|e| self.sema.type_of_expr(&e)).flatten();
+
+                            (ty, None)
+                        },
+                        ast::IdentPat(it) => {
+                            cov_mark::hit!(expected_type_if_let_with_leading_char);
+                            cov_mark::hit!(expected_type_if_let_without_leading_char);
+                            cov_mark::hit!(expected_type_match_arm_with_leading_char);
```

Comment:
I am a bit wary about several `hit!`s at the same place. To my mind, the difference between zero and one mark is huge, because that lets you to quickly identify the interesting test. Additional marks generally don't add as much value (other tests are probably nearby anyway), but they make the code more verbose.

Reply:
Yeah, I wasn't sure about this. I went forward with it because:

1. These marks were genuinely useful in helping me reason about this code during development, and I figure if they are useful now as I am writing this code they will be at least 2x as useful when myself or someone else needs to change it in the future. 
2. I considered perhaps only having marks on some of the tests but then I'd need to arbitrarily draw the line somewhere, which seems harder than putting them in each test. 

That said, I can remove them if you prefer that. Although if we want to remove them, I'd prefer to remove them all rather than only some of them (as discussed in point 2 above).

Reply:
I am fine with keeping them all! I'd be strongly against removing all, as having just one mark is hugely helpful. 

In the future, if someone sees this logic, they might wonder "ok, what's the actual rust code that causes us to hit this case?" By grepping for a mark, they can quickly identify one example.

Anyway, that's not really all that important

bors r+